### PR TITLE
soc: silabs_exx32: Make HFRCO frequency configurable

### PR DESCRIPTION
**Pull request by pkaranjekar**
This commit adds an option to configure the frequency
in Hz for the HFRCO clock source. The default value
for this option is 0 which skips the configuration of the HFRCO.
This feature is supported for efm32gg, efm32jg, efm32pg, efm32fg
and efm32mg SOCs currently.

This has been tested on the EFR32MG-SLTB004A board with the 
samples/basic/blinky application, verified with the debugger that the
configured frequency is actually set.

Signed-off-by: Pooja Karanjekar <pooja.karanjekar@lemonbeat.com>

### CODE REVIEW

**Review by chrta — APPROVED** [review]
I also tested this on the efm32pg board with different HFRCO frequency settings.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
can you rebase this PR on master, that should fix the doc issue.

**Review by MaureenHelm — APPROVED** [review]

**Comment by pkaranjekar** [comment]
Ok thanks, rebase done.

File: soc/arm/silabs_exx32/common/soc.c [hunk 1]
```diff
@@ -60,9 +60,16 @@ static ALWAYS_INLINE void clock_init(void)
...
 #elif (defined CONFIG_CMU_HFCLK_HFRCO)
 	/*
-	 * This is the default clock, the controller starts with, so nothing to
-	 * do here.
+	 * This is the default clock, the controller starts with
 	 */
+
+	if (CONFIG_CMU_HFRCO_FREQ) {
+		/* Setting system HFRCO frequency */
+		CMU_HFRCOBandSet(CONFIG_CMU_HFRCO_FREQ);
```
Comment (mnkp) [inline soc/arm/silabs_exx32/common/soc.c]: This call will not work on efm32wg and efm32hg series. The function is declared as
```
void                  CMU_HFRCOBandSet(CMU_HFRCOBand_TypeDef band);
```
where 
```
typedef enum {
  cmuHFRCOBand_1MHz  = 0,      /**< 1 MHz HFRCO band  */
  cmuHFRCOBand_7MHz  = 1,      /**< 7 MHz HFRCO band  */
  cmuHFRCOBand_11MHz = 2,     /**< 11 MHz HFRCO band */
  cmuHFRCOBand_14MHz = 3,     /**< 14 MHz HFRCO band */
  cmuHFRCOBand_21MHz = 4,     /**< 21 MHz HFRCO band */
} CMU_HFRCOBand_TypeDef;
```
Also, in this case `0` is a valid value.

**Review by mnkp — CHANGES_REQUESTED** [review]
Setting HFRCO frequency is a useful feature but unfortunately this implementation doesn't work on efm32wg and efm32hg series. I'm actually not sure what's the best way to handle this issue. I believe it would be also OK not to support configurable HFRCO frequency on efm32wg and efm32hg. It can be added later. But in this case CMU_HFRCO_FREQ Kconfig symbol shouldn't be visible for efm32wg and efm32hg series. The implementation in `soc.c` also would need to be modified a bit.

**Comment by pkaranjekar** [comment]
Yes, you are right. I have added the necessary changes.

File: soc/arm/silabs_exx32/Kconfig [hunk 2]
```diff
@@ -182,4 +193,12 @@ config SOC_GECKO_HAS_ERRATA_RTCC_E201
 	  feature enabled by setting CCV1TOP in RTCC_CTRL fails to wrap the counter
 	  when RTCC_CNT is equal to RTCC_CC1_CCV, as intended."
 
+config SOC_GECKO_HAS_HFRCO_FREQ
+	bool
+	help
+	  If enabled, indicates that configuration of HFRCO frequency is supported for this SOC.
```
Comment (mnkp) [inline soc/arm/silabs_exx32/Kconfig]: As is, the comment is misleading. The efm32hg, efm32wg series also support configuration of HFRCO frequency, however use a different method. What about saying:

If enabled, indicates that configuration of HFRCO frequency for this SOC is supported via FREQRANGE field.

If we add in the future support for  efm32hg, efm32wg series we would need to use another Kconfig option. This one should be reserved for SoCs that rely on FREQRANGE field. We could also rename it to `SOC_GECKO_HAS_HFRCO_FREQRANGE`.

**Review by mnkp — COMMENTED** [review]
Thanks, implementation looks good. I have a minor comment though.

File: soc/arm/silabs_exx32/Kconfig (hunk 2 above)
Comment (pkaranjekar) [reply soc/arm/silabs_exx32/Kconfig]: Done.

**Review by mnkp — APPROVED** [review]

**Pull request merged**